eclipse-theia/theia
# Support for vscode `l10n` API

#### What it does

Closes https://github.com/eclipse-theia/theia/issues/11893

VSCode 1.73 not only brought the new `l10n` namespace with it, but also changed the way language packs for vscode built-ins work. Theia now supports both _old_ and _new_ language packs and extension localization APIs (`l10n` and `vscode-nls`).

#### How to test

You will need 4 extensions to test this feature:
* [l10n-test](https://github.com/eclipse-theia/theia/files/10767190/l10n-test-0.0.1.zip) ([source here](https://github.com/msujew/vscode-l10n-test))
* [German language pack@1.75](https://github.com/eclipse-theia/theia/files/10767197/vscode-language-pack-de-1.75.2023021509.zip)
* [css@1.75](https://github.com/eclipse-theia/theia/files/10767205/css-1.0.0.zip)
* [css-language-features@1.75](https://github.com/eclipse-theia/theia/files/10839511/css-language-features-1.0.1.zip)

1. Install `l10n-test` and execute the `l10n:*` commands.
    1. The `Hello World` command will create a notification with `"Hello World!"`
    2. The `Show URI` command will show `UNDEFINED`
    3. The `Show Bundle` command will also show undefined.
2. Install the German language pack and change the locale to `German/de`
3. Repeat step 1 (search for `l10n`, the command names are localized)
    1. The `Hello World` command will create a notification with `"Hallo Welt!"`
    2. The `Show URI` command will show the file uri to the German language bundle of the `l10n-test` extension
    3. The `Show Bundle` command will display a JSON containing the language bundle content
4. Delete the builtin `css-language-features` extension
5. Install the provided `css` and `css-language-features` extensions.
6. Open a CSS file, and hover over a selector. It should show `Selektorspezifität` (German) instead of `Selector Specificity` (English) in the hover. This step asserts that extensions can still be localized using language packs.

Also make sure to test for regressions with existing language pack functionality:

1. Confirm that old (< 1.73 API version) language packs still work as expected.
2. Confirm that this also works in conjunction with localized built-in extensions (such as the `typescript-language-features` extension)
3. Deinstalling language packs and consequently reloading the application should actually delete all contributed localizations/language pack bundles.

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
